### PR TITLE
add KV_RW support

### PR DESCRIPTION
**Pull request by hannesm**
this was developed by @reynir and myself. I just added the last commit to use a single BLOCK.write operation. What is still missing is support if the sector_size is not 512!

### CODE REVIEW

**Comment by hannesm** [comment]
The CI failures:
- on macos, the command `truncate` is not present (any clue how to deal with this? should we add a depext?) anyone with macos has a clue about that basic unix utility?
- on windows, the expect test (in `test` subdirectory) does not work -- I suspect the `echo "Hello World" > archive/foo` does not work as intended -- how should be dealt with this (/cc @dinosaure who contributed that test)

**Comment by reynir** [comment]
Likely we need a write lock as well

**Comment by hannesm** [comment]
> Likely we need a write lock as well

indeed, added in [38c0b6d](https://github.com/mirage/ocaml-tar/pull/93/commits/38c0b6dc3a350c0f06b31c62cb5f704fe2dddc68)

**Comment by reynir** [comment]
Thanks! I pushed a small change that moves the lock from the module to the device. As stated in the commit the same module may be reused for different devices, e.g. when using mirage-block-partition. In that case we'd like to avoid a write in one partition blocking a write in another partition.

**Comment by hannesm** [comment]
I rebased on main, and added a PCLOCK functor argument to Make_KV_RW  (for the `mod_time`)

**Comment by dinosaure** [comment]
Hmmhmm, on test the `truncate` seems missing on Mac OS. I'm not sure who introduce these tests, probably @MisterDA. If it's the case, do you know how to properly fix it? Otherwise, this PR seems good but you probably want to experiment more @hannesm / @reynir ?

**Comment by hannesm** [comment]
I have no clue about how to fix the macos test (how to get truncate on macos).

I am working on a commit to support block sizes <> 512 for writing, once that is done, imho it is ready to be released.

**Comment by MisterDA** [comment]
> on macos, the command truncate is not present (any clue how to deal with this? should we add a depext?) anyone with macos has a clue how to extend a file? is there an alternative utility that works on all platforms?

macOS doesn't have the truncate command (from coreutils) but it's available with brew. It does have the C lib `truncate` function, you could write an extremely simple overlay, or disable that test on macOS.

> Hmmhmm, on test the truncate seems missing on Mac OS. I'm not sure who introduce these tests, probably @MisterDA. If it's the case, do you know how to properly fix it?

The test using truncate was introduced with this PR.

> on windows, the expect test (in test subdirectory) does not work -- I suspect the echo "Hello World" > archive/foo does not work as intended -- how should be dealt with this (/cc @dinosaure who contributed that test)

Has it been fixed? The CI doesn't show the failure.

**Comment by hannesm** [comment]
This looks fine now, apart from OpenSUSE failures:- 
Failure("KV_RO.read: read: End_of_file at file /tmp/build_8e2190_dune/tar-test2034e7-builder#05.tar offset 61440 with buffers of length [ 4096, 4096, 4096 ]")
- Failure("KV_RO.read: Cannot find the key /\"barf\"")

Maybe @reynir (who had some magic to fix earlier OpenSUSE issues) has an idea how to fix these?

**Comment by hannesm** [comment]
hmm, and the windows CI is failing, I fail to understand why... the `ENOENT` shouldn't occur (we do "convert_path" on Windows)... any insights, anyone?

**Comment by hannesm** [comment]
ok, I think now lots of issues are fixed, the read and write were reading over the end of the file... fixed in [e90ed6d](https://github.com/mirage/ocaml-tar/pull/93/commits/e90ed6d8df88627f19b1b4db151a83f0e06383f8)

File: mirage/tar_mirage.ml [hunk 1]
```diff
@@ -297,14 +296,21 @@ module Make_KV_RW (CLOCK : Mirage_clock.PCLOCK) (BLOCK : Mirage_block.S) = struc
     type 'a t = 'a Lwt.t
     let really_write out_channel data =
       assert (Cstruct.length data <= 512);
-      let data = Cstruct.(append data (create (length data - 512))) in
-      let sector = Int64.(div out_channel.offset (of_int out_channel.info.sector_size)) in
-      BLOCK.write out_channel.b sector [ data ] >>= function
+      let data = Cstruct.(append data (create (512 - length data))) in
+      let sector_size = out_channel.info.sector_size in
+      let sector = Int64.(div out_channel.offset (of_int sector_size)) in
+      let block = Cstruct.create sector_size in
```
[the remark this answers is not in the document]
Comment (hannesm) [reply mirage/tar_mirage.ml:302]: yes, I agree we should assert that explicitly. and indeed, it is both RO and RW.

File: mirage/tar_mirage.ml [hunk 2]
```diff
@@ -333,38 +339,67 @@ module Make_KV_RW (CLOCK : Mirage_clock.PCLOCK) (BLOCK : Mirage_block.S) = struc
...
+              (Cstruct.length data)
+              (to_int (sub sector_size start_sector_offset))
           in
-          match blocks with
-          | [] -> [], []
-          | hd :: tl -> [ hd ], tl
+          Cstruct.split data s
+        in
+        let remaining_blocks, last_block =
+          let blocks = to_int (sub data_sectors (if end_sector_offset = 0L then 1L else 2L)) in
```
[the remark this answers is not in the document]
Comment (hannesm) [reply mirage/tar_mirage.ml:362]: yes, it as well confuses me... but I'm not sure how to express it differently: if the last byte to-be-written is at the end of a block (last byte), the data_sectors will be the next one and the end_sector_offset 0. :/ I'll try to come up with nicer code.

**Comment by hannesm** [comment]
Thanks for your review, @reynir. I pushed individual commits that address your comments. Please let me know whether that is good now :)

File: mirage/tar_mirage.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (reynir) [reply mirage/tar_mirage.ml]: Sorry, my comment was not very constructive. I meant to expand on it before publishing the review. What confused me was why we were doing this, and a comment explaining the problem could have been enough. But I like changes in 53a299d8a8f28bb18ce2dbcfd28f320f1d58b532, particularly being more clear about blocks vs sectors and that we're interested in "full" sectors vs potential partial sector at the end, and I feel less confused why we're doing this :)

**Comment by reynir** [comment]
I tried to improve the errors:
- by preserving the errors in the header writer with a custom exception, and reraise any other exceptions,
- by preserving read errors in write operations, and add an explanation in `pp_write_error` that the read error occured during a write
I also documented the exception in `connect` and some internal invariants in `insert` and `update_insert`.

I think it is good to merge now.

**Review by dinosaure — APPROVED** [review]

**Comment by hannesm** [comment]
I think this looks good. We could eventually implement `rename`, but that can as well be done in a separate PR later.

**Comment by hannesm** [comment]
thanks @reynir for the additional cleanups

File: mirage/tar_mirage.ml [hunk 3]
```diff
@@ -233,9 +237,200 @@ module Make_KV_RO (BLOCK : Mirage_block.S) = struct
...
+        Lwt.return (Ok ()))
+
+  let remove _ _ =
+    Lwt.return (Error `Append_only)
+
+  let rename _ ~source:_ ~dest:_ =
+    Lwt.return (Error `Append_only)
+
+  let set_partial _ _ ~offset:_ _ =
+    Lwt.return (Error `Append_only)
```
[the remark this answers is not in the document]
Comment (hannesm) [reply mirage/tar_mirage.ml:432]: Well, if you're using mirage-kv 5.0.0, they are required & exported :)

[the remark this answers is not in the document]
Comment (dinosaure) [reply mirage/tar_mirage.ml:432]: I think it's needed for the last `Mirage_kv.S` interface (`5.0.0`). We probably, as I did elsewhere (`git-kv`), put `(flags (-w -32))` to let `tar-mirage` usable with `mirage-kv.5.0.0` and `4.0.0`.

[the remark this answers is not in the document]
Comment (MisterDA) [reply mirage/tar_mirage.ml:432]: My bad, thanks to both ;)

**Pull request merged**